internal/compiler/parser/type.rs

```diff
@@ -85,3 +85,61 @@ pub fn parse_struct_declaration(p: &mut impl Parser) -> bool {
     parse_type_object(&mut *p);
     true
 }
+
+pub fn parse_rustattr(p: &mut impl Parser) -> bool {
+    let checkpoint = p.checkpoint();
+    debug_assert_eq!(p.peek().as_str(), "@");
+    p.consume(); // "@"
+    if p.peek().as_str() != "rust-attr" {
+        p.expect(SyntaxKind::AtRustAttr);
+    }
+    p.consume(); // "rust-attr"
+    p.expect(SyntaxKind::LParent);
+    parse_parentheses(&mut *p);
+    if p.peek().as_str() == "export" {
+        p.consume();
+    }
+    let mut p = p.start_node_at(checkpoint, SyntaxKind::StructDeclaration);
+    p.consume(); // "struct"
+    {
+        let mut p = p.start_node(SyntaxKind::DeclaredIdentifier);
+        p.expect(SyntaxKind::Identifier);
+    }
+
+    if p.peek().kind() == SyntaxKind::ColonEqual {
+        p.warning("':=' to declare a struct is deprecated. Remove the ':='");
+        p.consume();
+    }
+
+    parse_type_object(&mut *p);
+    true
+}
+
+fn parse_parentheses(p: &mut impl Parser) -> bool {
+    let mut p = p.start_node(SyntaxKind::AtRustAttr);
+    let mut opened = 0;
+    let mut closed = 0;
+    while closed <= opened {
+        if p.peek().kind() == SyntaxKind::LParent {
+            opened += 1;
+        }
+        if p.peek().kind() == SyntaxKind::RParent {
+            closed += 1;
+        }
+        if closed == opened
+            && opened != 0
+            && closed != 0
+            && p.peek().kind() != SyntaxKind::Comma
```

Comment:
Why do you check for a coma?

Note, you can also see https://github.com/slint-ui/slint/blob/0c64312383bfea39c4b01d518619bbde2c2c7062/tools/lsp/server_loop.rs#L901  for some example of code that tries to find balanced parentheses

Reply:
I suppose it will parse if check comma between element of rust_attr . as we have extra RPRANT which is for rust_attr e can use it . but if we want to check for all possibilty in rust-attr it worth to validate as rust too and you mentioned e can just check it out . it will be big a problem to return error or annouce which element missed . as we want to relay on rust compile we relay on it for this .